### PR TITLE
Resolved issue for high precision MLE estimation

### PR DESCRIPTION
**Pull request by LucaCappelletti94**
* Data type used for multiplicities in MLE estimation for HLL counters were insufficient for high precisions (16, 17, 18)
* Resolves issue #3295 
* Introduced new trait `MultiplicityInteger` to handle cleanly without code duplication and the dispatching of different data types given the requested precision at runtime, without introducing marker structs or other distinctive struct attributes. Since the MLE estimation is rather slow, adding an if chain before calling the optimal variant does not add any significant time overhead.

### CODE REVIEW

**Comment by LucaCappelletti94** [comment]
@luizirber do check out also this PR please, being an actual bug I believe it is more important

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 1]
```diff
@@ -1,38 +1,91 @@
...
+    cmp,
+    ops::{Add, AddAssign, Shl, Sub, SubAssign},
+};
 
 pub type CounterType = u8;
 
-pub fn counts(registers: &[CounterType], q: usize) -> Vec<u16> {
-    let mut counts = vec![0; q + 2];
+/// Trait for types that can be used as multiplicity integers.
+pub trait MulteplicityInteger:
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
pub trait MultiplicityInteger:
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 2]
```diff
@@ -1,38 +1,91 @@
...
+    /// The zero value.
+    const ZERO: Self;
+    /// The one value.
+    const ONE: Self;
+
+    /// Convert the value to a `f64`.
+    fn to_f64(self) -> f64;
+}
+
+macro_rules! impl_multeplicity_integer {
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
macro_rules! impl_multiplicity_integer {
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 3]
```diff
@@ -1,38 +1,91 @@
...
+    ($($t:ty),*) => {
+        $(
+            impl MulteplicityInteger for $t {
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
            impl MultiplicityInteger for $t {
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 4]
```diff
@@ -1,38 +1,91 @@
...
+
+                fn to_f64(self) -> f64 {
+                    self as f64
+                }
+            }
+        )*
+    };
+}
+
+impl_multeplicity_integer!(u8, u16, u32);
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
impl_multiplicity_integer!(u8, u16, u32);
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 5]
```diff
@@ -1,38 +1,91 @@
...
+
+pub fn counts<M: MulteplicityInteger>(registers: &[CounterType], q: usize) -> Vec<M> {
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
pub fn counts<M: MultiplicityInteger>(registers: &[CounterType], q: usize) -> Vec<M> {
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 6]
```diff
@@ -1,38 +1,91 @@
...
+        counts[*k as usize] += M::ONE;
     }
 
     counts
 }
 
 #[allow(clippy::many_single_char_names)]
-pub fn mle(counts: &[u16], p: usize, q: usize, relerr: f64) -> f64 {
-    let m = 1 << p;
+pub fn mle<M: MulteplicityInteger>(counts: &[M], p: usize, q: usize, relerr: f64) -> f64 {
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
pub fn mle<M: MultiplicityInteger>(counts: &[M], p: usize, q: usize, relerr: f64) -> f64 {
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 7]
```diff
@@ -111,57 +164,82 @@ pub fn joint_mle(
...
+    } else {
+        assert!(p == 16 || p == 17 || p == 18);
+        joint_mle_dispatch::<u32>(k1, k2, p, q)
+    }
+}
+
+/// Calculate the joint maximum likelihood of A and B.
+///
+/// Returns a tuple (only in A, only in B, intersection)
+fn joint_mle_dispatch<M: MulteplicityInteger>(
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
fn joint_mle_dispatch<M: MultiplicityInteger>(
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 8]
```diff
@@ -111,57 +164,82 @@ pub fn joint_mle(
...
-    counts_axb_half[q] = k1.len() as u16;
-    counts_bxa_half[q] = k2.len() as u16;
+    counts_axb_half[q] = M::try_from(k1.len()).unwrap();
+    counts_bxa_half[q] = M::try_from(k2.len()).unwrap();
 
     for _q in 0..q {
         counts_axb_half[_q] = cg1[_q] + ceq[_q] + cg2[_q + 1];
         debug_assert!(counts_axb_half[q] >= counts_axb_half[_q]);
-        counts_axb_half[q] -= counts_axb_half[_q];
+        let multeplicity_q = counts_axb_half[_q];
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
        let multiplicity_q = counts_axb_half[_q];
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 9]
```diff
@@ -111,57 +164,82 @@ pub fn joint_mle(
...
+        counts_axb_half[q] -= multeplicity_q;
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
        counts_axb_half[q] -= multiplicity_q;
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 10]
```diff
@@ -111,57 +164,82 @@ pub fn joint_mle(
...
 
         counts_bxa_half[_q] = cg2[_q] + ceq[_q] + cg1[_q + 1];
         debug_assert!(counts_bxa_half[q] >= counts_bxa_half[_q]);
-        counts_bxa_half[q] -= counts_bxa_half[_q];
+        let multeplicity_q = counts_bxa_half[_q];
+        counts_bxa_half[q] -= multeplicity_q;
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
        counts_bxa_half[q] -= multiplicity_q;
```

File: src/core/src/sketch/hyperloglog/estimators.rs [hunk 11]
```diff
@@ -111,57 +164,82 @@ pub fn joint_mle(
...
         counts_axb_half[_q] = cg1[_q] + ceq[_q] + cg2[_q + 1];
         debug_assert!(counts_axb_half[q] >= counts_axb_half[_q]);
-        counts_axb_half[q] -= counts_axb_half[_q];
+        let multeplicity_q = counts_axb_half[_q];
+        counts_axb_half[q] -= multeplicity_q;
 
         counts_bxa_half[_q] = cg2[_q] + ceq[_q] + cg1[_q + 1];
         debug_assert!(counts_bxa_half[q] >= counts_bxa_half[_q]);
-        counts_bxa_half[q] -= counts_bxa_half[_q];
+        let multeplicity_q = counts_bxa_half[_q];
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/estimators.rs]:
```suggestion
        let multiplicity_q = counts_bxa_half[_q];
```

File: src/core/src/sketch/hyperloglog/mod.rs [hunk 12]
```diff
@@ -374,4 +396,53 @@ mod test {
         assert_eq!(hll_new.registers, hll.registers);
         assert_eq!(hll_new.ksize, hll.ksize);
     }
+
+    #[test]
+    /// Test to cover corner cases in the MLE calculation
+    /// that may happen at resolutions 16, 17 or 18, i.e.
+    /// cases with 2^16 == 65536, 2^17 == 131072, 2^18 == 262144.
+    ///
+    /// In such cases, the MLE multeplicities which were earlier
```
Comment (luizirber) [inline src/core/src/sketch/hyperloglog/mod.rs]:
```suggestion
    /// In such cases, the MLE multiplicities which were earlier
```

**Review by luizirber — APPROVED** [review]
Very elegant solution, thanks @LucaCappelletti94!
